### DISCUSSION TITLE
Used addresses

### DISCUSSION THREAD

**Discussion by szist** [CardanoSolutions/kupo#78] [Ideas]
#### Describe your idea, in simple words.
Ability to filter out full addresses based on either delegation or payment credential.

#### Why is it a good idea?
It could help identifying which addresses are tied to a script (e.g. DEX pool addresses where each pool has a different delegation) or getting a list of addresses from where the delegator ADA is considered.
Secondly, it would help with interacting with HD wallets to detect which addresses were used already. 
The sqlite structure and indexes would already help with it and some distinct select on top of the addresses should solve this.

#### Are you willing to work on it yourself?
Maybe, but I'm not as fluent in Haskell.

**Comment by KtorZ** [comment]
> Ability to filter out full addresses based on either delegation or payment credential.

Hey, I've got good news, you can already do that 😄! 
The `GET /match/{pattern}` endpoint accepts any kind of _patterns_, including patterns to match on either delegation or payment credentials. Have a look at the [user manual on Patterns](https://cardanosolutions.github.io/kupo/#section/Patterns) for more details about the syntax and how to construct patterns.

>  Secondly, it would help with interacting with HD wallets to detect which addresses were used already.

There's already a field on each match result indicating whether a UTxO has been spent or not as well as a query flag that allows you to filter UTxO based on that criteria. Surely, querying any pattern and requesting only the `?spent` one would give you only results that have already been used.

**Reply by szist** [reply]
But using this would mean that I would still need to post-process on the client side all the utxos that are received (all spent utxos), which could be computationally quite heavy. I think it would be much more efficient to get a list of these addresses directly from sqlite. But I can understand if this would be out of scope for the project if it only wants to focus on utxos :)

**Reply by KtorZ** [reply]
Yes indeed, Kupo comes with a relatively flexible API but it's also meant to be general purpose and not a backend for **a specific** client; Having said that, I am not quite sure what you mean by "post-process" here, given that the API would already return you a list of matches containing the addresses you're looking for? 

Are you asking for the API to return _only the addresses_ without other fields? In which case, I am afraid that this is more of shortcomings that comes from defining a RESTful interface rather than something like GraphQL which enables this kind of patterns. REST Is "simple" and works well when you spend some times in designing the API properly (which I hope I did), but it also often tends to return more data than what your specific use-case demand. 

Note that: 

(1) It's only about traversing a list here, so the work is in O(n) of the size of the results, which should be manageable really. 
(2) The server streams the results to clients, one result per line, so in principle, you can leverage that to start post-processing concurrently as soon as you get the first result, such that when the server is done sending the response, you're also done post-processing.

**Reply by szist** [reply]
All true, post processing is fine as long as the `n` is a relatively low number. For example once you have active trading accounts and n grows, it gets increasingly costly to post-process the txos.

If you also consider paging, then the list might become incomplete, unless the client fetches everything. And yeah I meant just listing the addresses used with an sql distinct select. It seemed like a low hanging fruit :) an alternative to https://docs.blockfrost.io/#tag/Cardano-Accounts/paths/~1accounts~1{stake_address}~1mirs/get on delegation key side, and no alternative for getting all addresses associated with a scripthash currently :)

**Reply by KtorZ** [reply]
To be frank, I don't think post processing a list with a single traversal would be a major issue. I am constantly benchmarking against the jpeg.store or SundaeSwap contract addresses which have 100k+ UTxOs. Fetching, streaming and post processing the whole result (150MB+ of data) with jq takes less than 3s. And that's about the biggest request one can do on mainnet at the moment.

Regarding pagination, you're 100% right. I'll drop some thoughts on the other discussion tomorrow.

**Reply by szist** [reply]
Well, if you want use locally 150m is not that much. Having microservices or using it to send data over mobile network, that's quite a lot of data. :)

**Reply by KtorZ** [reply]
But Kupo is primarily meant to be used as a local service 😁...

Maybe you could tell more about your use-case?

**Reply by szist** [reply]
I have a few in mind :D. But I would just leave it for later or to a dm. For one of the first specific case I am looking to replace that exact endpoint I posted there from blockfrost. I don't like relying on such external centralized services. Carp seems too heavy (even though it tries to be modular) , db-sync-based stuff (even just with ledger state not much better).

(second more of a painpoint I can share, blockfrost and explorers lack the ability to get the addresses for a script, which already caused me a few issues elsewhere, as well for some aggregators trying to measure TVL and assuming that all orders/lps/farms would be on the same address not considering that all those utxos could be delegated. There you could have a lighter kupo running to get all addresses for a script and the use blockfrost/other address based service to get the rest )
